### PR TITLE
Verify services pages are displayed correctly

### PR DESCRIPTION
**Pull request by ThaoThaoTran**
This PR is to verify all links and pages are displayed correctly when navigating to it

### CODE REVIEW

File: src/Codelink_Website/core/testcases/ServicesTest.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test
+    public void Services_Links_And_Pages(Method method){
+        ExtentTestManager.startTest(method.getName(),"Home Page");
+
+        log.info("Select Services home bar");
+        homePage.selectHomeBar("Services");
+
+        logInfo("Select Artificial Intelligence & Machine Learning");
+        homePage.selectItemInHomeBar("Artificial Intelligence & Machine Learning");
```
Comment (chiivv) [inline src/Codelink_Website/core/testcases/ServicesTest.java]: [CLARIFICATION] What is the rationale behind having distinct logging mechanisms here, `log.info` and `logInfo`?

File: src/Codelink_Website/core/pageObjects/HomePage.java [hunk 2]
```diff
@@ -16,4 +17,19 @@ public ContactUsPage navigateToContactUsPage() {
         clickToElement(driver, HomePageUI.CONTACT_US_BTN);
         return new ContactUsPage(driver);
     }
+
+    public void selectHomeBar(String homeBar) {
+        waitForElementVisible(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR,homeBar);
+        hoverMouseToElement(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR,homeBar);
+    }
```
Comment (chiivv) [inline src/Codelink_Website/core/pageObjects/HomePage.java]:
```suggestion
    public void selectHomeBar(String homeBar) {
        waitForElementVisible(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR, homeBar);
        hoverMouseToElement(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR, homeBar);
    }
```

**Review by chiivv — CHANGES_REQUESTED** [review]

File: src/Codelink_Website/core/testcases/ServicesTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        log.info("Select Services home bar");
+        homePage.selectHomeBar("Services");
+
+        logInfo("Select Software Development");
+        homePage.selectItemInHomeBar("Software Development");
+
+        logInfo("Verify Software Development page is displayed correctly");
+        Assert.assertTrue(homePage.isTitleDisplayed("Software Development"));
+        Assert.assertEquals(homePage.getPageURL(driver), GlobalConstants.SOFTWARE_DEVELOPMENT_PAGE);
```
Comment (chiivv) [inline src/Codelink_Website/core/testcases/ServicesTest.java]: [BLOCKER] I have a concern regarding the test case design and page structure. It seems we are leaning towards adopting the Page Object Model. If that's the case, there might be additional considerations we need to address
- Add new pages under src/Codelink_Website/core/pageObjects
```
public class softwareDevelopmentPage
    public void verifyTitle();
    public void verifyUrl();
```
- Update validations in test case
```
homePage.selectItemInHomeBar("Software Development");
softwareDevelopmentPage.waitForPageLoad();
softwareDevelopmentPage.verifyTitle();
softwareDevelopmentPage.verifyUrl();

```

File: src/Codelink_Website/core/interfaces/ServicesUI.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package interfaces;
+
+public class ServicesUI {
+    public static final String DYNAMIC_BUTTON_HOME_BAR="xpath=//div[@class='HeaderV2-module--right-container-desktop--2muX7']//span[text()='%s']";
+    public static final String DYNAMIC_BUTTON_SELECT_HOME_BAR="xpath=//div[@class='DesktopMenuExtension-module--list--i50hg']//h6[text()='%s']";
```
Comment (chiivv) [inline src/Codelink_Website/core/interfaces/ServicesUI.java]: [BLOCKER] Looks like they are dynamic web elements. We might need to rely on the fixed part in class name, e.g:  `xpath=//div[contains(@class, 'HeaderV2-module--right-container-desktop')]//span[text()='%s']`

File: src/Codelink_Website/core/pageObjects/CommonFunctionsPage.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    public void selectHomeBar(String homeBar) {
+        waitForElementVisible(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR, homeBar);
+        hoverMouseToElement(driver, ServicesUI.DYNAMIC_BUTTON_HOME_BAR, homeBar);
+    }
+
+    public void selectItemInHomeBar(String item) {
+        waitForElementVisible(driver, ServicesUI.DYNAMIC_BUTTON_SELECT_HOME_BAR, item);
+        clickToElement(driver, ServicesUI.DYNAMIC_BUTTON_SELECT_HOME_BAR, item);
+    }
```
Comment (chiivv) [inline src/Codelink_Website/core/pageObjects/CommonFunctionsPage.java]: [SUGGESTION] These functions seem specifically applicable to the Home page. If that's the case, please move them to the Home page

File: src/Codelink_Website/core/testcases/ServicesTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        logInfo("Select Services in home bar");
+        commonFunctionsPage.selectHomeBar("Services");
+
+        logInfo("Open Artificial Intelligence & Machine Learning page");
+        commonFunctionsPage.selectItemInHomeBar("Artificial Intelligence & Machine Learning");
+
+        logInfo("Verify Artificial Intelligence & Machine Learning page is displayed correctly");
+        Assert.assertTrue(commonFunctionsPage.isTitleDisplayed("Artificial Intelligence & Machine Learning"));
+        Assert.assertEquals(commonFunctionsPage.getPageURL(driver), GlobalConstants.ARTIFICIAL_INTELLIGENCE_PAGE);
```
Comment (chiivv) [inline src/Codelink_Website/core/testcases/ServicesTest.java]: [BLOCKER] commonFunctionsPage is not an actual page. The best practices should be creating separate pages that I mentioned in the previous review. https://github.com/CodeLinkIO/codelink-website-automation/pull/4#discussion_r1446976697

**Review by chiivv — CHANGES_REQUESTED** [review]

**Review by chiivv — APPROVED** [review]

**Review by PhuongMaiAutomation — APPROVED** [review]

**Pull request merged**